core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -299,41 +300,55 @@ public boolean topicExists(final String topic) throws TopicRepositoryException {
     public void syncPostBatch(
             final String topicId, final List<BatchItem> batch, final String eventType, final boolean delete)
             throws EventPublishingException {
+
         final Producer<byte[], byte[]> producer = kafkaFactory.takeProducer();
         try {
-            final Map<String, String> partitionToBroker = producer.partitionsFor(topicId).stream()
-                    .filter(partitionInfo -> partitionInfo.leader() != null)
-                    .collect(
-                            Collectors.toMap(
-                                    p -> String.valueOf(p.partition()),
-                                    p -> p.leader().idString() + "_" + p.leader().host()));
             batch.forEach(item -> {
                 Preconditions.checkNotNull(
                         item.getPartition(), "BatchItem partition can't be null at the moment of publishing!");
-                item.setBrokerId(partitionToBroker.get(item.getPartition()));
             });
 
-            final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
+            final Function<BatchItem, CompletableFuture<Exception>> sendItem =
+                    item -> publishItem(producer, topicId, eventType, item, delete);
+
+            final List<CompletableFuture<Exception>> futures = new LinkedList<>();
+            final Map<String, CompletableFuture<Exception>> futuresByKey = new HashMap<>();
+
             for (final BatchItem item : batch) {
-                final String brokerId = item.getBrokerId();
-                if (brokerId == null) {
-                    item.updateStatusAndDetail(EventPublishingStatus.FAILED,
-                            String.format("No leader for partition: %s, topic: %s.", item.getPartition(), topicId));
-                    LOG.error("Failed to publish to kafka. No leader for ({}:{}).",
-                            topicId, item.getPartition());
-                    continue;
+                final String itemKey = item.getEventKey();
```

Comment:
it feels like that logic should be one level higher and the grouping should be done not by the event id, but rather per partition, this is what matters for publishing. It sounds logically more correct. probably, it will change the method signature 🤷